File: momentra/momentra/Shell/PersonalEmpty/PersonalLifestyleDerived.swift
import Foundation

/// Lifestyle Vitality Index + axis scores from pulse + widgetPayload (Figma 505:12365).
enum PersonalLifestyleDerived {
    struct AxisScores {
        let joy: String
        let fulfillment: String
        let vitality: String
        let exploration: String
    }

    static func payloadNumber(_ payload: [String: AnyDecodable]?, key: String) -> Double? {
        guard let raw = payload?[key]?.value else { return nil }
        if let n = raw as? Double { return n }
        if let n = raw as? Int { return Double(n) }
        if let s = raw as? String { return PersonalLifeOpsDerived.scoreNumber(s) }
        if let n = raw as? NSNumber { return n.doubleValue }
        return nil
    }

    static func displayFromPayloadOrFallback(
        payload: [String: AnyDecodable]?,
        payloadKey: String,
        fallback: String?
    ) -> String {
        if let n = payloadNumber(payload, key: payloadKey) {
            return String(Int(n.rounded()))
        }
        return PersonalLifeOpsDerived.displayScore(fallback)
    }

    static func axisScores(pulse: APIClient.PersonalPulsePayload?) -> AxisScores {
        let payload = pulse?.widgetPayload
        return AxisScores(
            joy: displayFromPayloadOrFallback(payload: payload, payloadKey: "joyScore", fallback: pulse?.recoveryScore),
            fulfillment: displayFromPayloadOrFallback(payload: payload, payloadKey: "fulfillmentScore", fallback: pulse?.wellbeingScore),
            vitality: displayFromPayloadOrFallback(payload: payload, payloadKey: "vitalityScore", fallback: pulse?.rhythmScore),
            exploration: displayFromPayloadOrFallback(
                payload: payload,
                payloadKey: "explorationScore",
                fallback: PersonalLifeOpsDerived.attentionDisplay(count: pulse?.attentionCount)
            )
        )
    }

    static func vitalityIndex(pulse: APIClient.PersonalPulsePayload?) -> Int? {
        let axes = axisScores(pulse: pulse)
        let nums = [axes.joy, axes.fulfillment, axes.vitality, axes.exploration]
            .compactMap { PersonalLifeOpsDerived.scoreNumber($0) }
        if nums.isEmpty {
            return PersonalLifeOpsDerived.scoreNumber(pulse?.wellbeingScore).map { Int($0.rounded()) }
        }
        return Int((nums.reduce(0, +) / Double(nums.count)).rounded())
    }

    static func vitalityIndexDisplay(pulse: APIClient.PersonalPulsePayload?) -> String {
        vitalityIndex(pulse: pulse).map(String.init) ?? "—"
    }

    static func networkStability(pulse: APIClient.PersonalPulsePayload?) -> String {
        guard let index = vitalityIndex(pulse: pulse) else { return "Awaiting first signals" }
        if index >= 75 { return "Flourishing" }
        if index >= 50 { return "Growing" }
        return "Building"
    }

    static func experienceCount(pulse: APIClient.PersonalPulsePayload?) -> Int {
        Int(payloadNumber(pulse?.widgetPayload, key: "experienceCount") ?? 0)
    }

    static func spendPairs(pulse: APIClient.PersonalPulsePayload?) -> [(String, String)] {
        guard let map = pulse?.widgetPayload?["spendByCurrency"]?.value as? [String: Any] else { return [] }
        return map.compactMap { key, value in
            let v = "\(value)"
            guard !key.isEmpty, !v.isEmpty else { return nil }
            return (key, v)
        }
    }
}

File: apk/app/src/main/java/com/example/momentra/ui/shell/personal/lifestyle/create/PersonalLifestyleDerived.kt
package com.example.momentra.ui.shell.personal.lifestyle.create

import com.example.momentra.data.api.PersonalPulseDto
import kotlin.math.roundToInt
import com.example.momentra.ui.shell.personal.lifeops.create.PersonalLifeOpsDerived

/** Lifestyle Vitality Index + axis scores from pulse widgetPayload (PX-2 precision refresh writes CANONICAL_COUNTS). */
object PersonalLifestyleDerived {
    data class AxisScores(
        val joy: String,
        val fulfillment: String,
        val vitality: String,
        val exploration: String,
    )

    fun payloadNumber(payload: Map<String, Any>?, key: String): Double? {
        val raw = payload?.get(key) ?: return null
        return when (raw) {
            is Number -> raw.toDouble()
            is String -> PersonalLifeOpsDerived.scoreNumber(raw)
            else -> null
        }
    }

    fun displayFromPayloadOrFallback(
        payload: Map<String, Any>?,
        payloadKey: String,
        fallback: String?,
    ): String {
        val fromPayload = payloadNumber(payload, payloadKey)
        if (fromPayload != null) return fromPayload.roundToInt().toString()
        return PersonalLifeOpsDerived.displayScore(fallback)
    }

    fun axisScores(pulse: PersonalPulseDto?): AxisScores {
        val payload = pulse?.widgetPayload
        return AxisScores(
            joy = displayFromPayloadOrFallback(payload, "joyScore", pulse?.recoveryScore),
            fulfillment = displayFromPayloadOrFallback(payload, "fulfillmentScore", pulse?.wellbeingScore),
            vitality = displayFromPayloadOrFallback(payload, "vitalityScore", pulse?.rhythmScore),
            exploration = displayFromPayloadOrFallback(
                payload,
                "explorationScore",
                PersonalLifeOpsDerived.attentionDisplay(pulse?.attentionCount),
            ),
        )
    }

    fun vitalityIndex(pulse: PersonalPulseDto?): Int? {
        val axes = axisScores(pulse)
        val nums = listOf(axes.joy, axes.fulfillment, axes.vitality, axes.exploration)
            .mapNotNull { PersonalLifeOpsDerived.scoreNumber(it) }
        if (nums.isEmpty()) {
            return PersonalLifeOpsDerived.scoreNumber(pulse?.wellbeingScore)?.roundToInt()
        }
        return (nums.average()).roundToInt()
    }

    fun vitalityIndexDisplay(pulse: PersonalPulseDto?): String {
        return vitalityIndex(pulse)?.toString() ?: "—"
    }

    fun networkStability(pulse: PersonalPulseDto?): String {
        val index = vitalityIndex(pulse) ?: return "Awaiting first signals"
        return when {
            index >= 75 -> "Flourishing"
            index >= 50 -> "Growing"
            else -> "Building"
        }
    }

    fun experienceCount(pulse: PersonalPulseDto?): Int {
        val raw = payloadNumber(pulse?.widgetPayload, "experienceCount") ?: return 0
        return raw.toInt()
    }

    fun spendPairs(pulse: PersonalPulseDto?): List<Pair<String, String>> =
        (pulse?.widgetPayload?.get("spendByCurrency") as? Map<*, *>)
            ?.entries
            ?.mapNotNull { (k, v) ->
                val key = k?.toString()?.takeIf { it.isNotBlank() } ?: return@mapNotNull null
                val value = v?.toString()?.takeIf { it.isNotBlank() } ?: return@mapNotNull null
                key to value
            }
            .orEmpty()

    fun momentumLabels(pulse: PersonalPulseDto?, activities: List<Pair<String, String>>): List<String> {
        val axes = axisScores(pulse)
        return listOf(
            if (axes.joy == "—") "Joy pending" else "Joy Rising",
            if (experienceCount(pulse) <= 0) "Ritual quiet" else "Experiences logged",
            if (pulse?.moodState.isNullOrBlank()) "Mood pending" else pulse?.moodState ?: "Mood pending",
            if (spendPairs(pulse).isEmpty()) "Budget quiet" else "Budget active",
        )
    }

    fun statusBadge(score: String, axis: String): String = PersonalLifeOpsDerived.statusBadge(score, axis)
}
